### ISSUE TITLE
Missing method exception

### ISSUE THREAD

**Issue by vsavic** [wwwlicious/servicestack-authentication-identityserver#14]
When I try to access protected api method, I got redirected to `/auth/IdentityServer?redirect=path-to-my-api-method` with following exception:

>[Authenticate: 2018-10-24 12:24:00 AM]: [REQUEST: {provider:IdentityServer}] System.MissingMethodException: Method not found: 'ServiceStack.Web.INameValueCollection ServiceStack.Web.IRequest.get_QueryString()'. at ServiceStack.Authentication.IdentityServer.Providers.UserAuthProvider.GetReferrerUrl(IServiceBase authService, IAuthSession session, Authenticate request) at ServiceStack.Authentication.IdentityServer.Providers.UserAuthProvider.Init(IServiceBase authService, IAuthSession& session, Authenticate request) at ServiceStack.Authentication.IdentityServer.Providers.UserAuthProvider.AuthenticateAsync(IServiceBase authService, IAuthSession session, Authenticate request)

I thought this would redirect me to identity server instance so I can login (locally or to use registered external providers), any idea why is this happening?

**Comment by stuartbfs** [comment]
Looks like you might be using a different version of Service Stack if the method is missing

**Comment by vsavic** [comment]
I'm using Service Stack version 5.2

**Comment by stuartbfs** [comment]
The plugin was built with version 4.0.56 so version 5 likely isn't supported.

**Comment by stuartbfs** [comment]
As a guess, you might need to set the tea version using ServicePointManager depending on the certificate being used.

e.g.

System.Net.ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12 | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls;

**Comment by vsavic** [comment]
Just wanted to confirm that this library can work with latest SS 5.4.1 and core 2.1, however I'm still struggling with popup which appear after I login into IdentityServer saying:
>The information you have entered on this page will be sent over an insecure connection and could be read by a third party.
>
>Are you sure you want to send this information?

Any idea if something might be off with client registration?

**Comment by stuartbfs** [comment]
Is your Service Stack application is using HTTP and your Identity Server instance is HTTPS? If so the browser will give you a warning.

**Comment by vsavic** [comment]
Yes, my app is behind the http, and identity server behind https, and warning is fine (I was receiving it with .net core (without your plugin as well, but just once), but the thing is this time this warning keeps triggering callback method, and from callback method in your plugin, and never gets redirected to my API method.

**Comment by vsavic** [comment]
This is the method where it keeps landing after I click continue over and over:

<img src="https://image.prntscr.com/image/yVrI--KTRnKC9O7-SVz8zg.png"/>

```lang=csharp
public override object Authenticate(IServiceBase authService, IAuthSession session, Authenticate request)
```

**Comment by stuartbfs** [comment]
It’s probably failing to authenticate on the Service Stack. You’ll need to confirm the callback url is configured correctly in identity server.

**Comment by vsavic** [comment]
In [this file](https://github.com/wwwlicious/servicestack-authentication-identityserver/blob/7685efe2f6ff2bfd9022e792911330383d603faa/src/ServiceStack.Authentication.IdentityServer/Providers/UserAuthProvider.cs#L204) from your this condition with IdentityServer4 this condition was never met:
```lang=csharp
(!httpRequest.QueryString["code"].IsNullOrEmpty())
```
...because this `code` was not transferred as part of URL query string, so I replaced it with:   
```lang=csharp
(!GetRequestValue("code", httpRequest).IsNullOrEmpty())
```

After that with my local identityserver4 everything works fine, however, with Azure I'm just having one more issue, and it's with `nonce`.

When enter this method from my azure instance: `IsValidIdToken`, `idAuthTokens.Nonce` is null, so somehow it got lost after it's set once in `AuthenticateClient` with:
```lang=csharp
 if (idAuthTokens != null)
            {                
                idAuthTokens.Nonce = nonce;
            }
```

>fail: ServiceStack.Authentication.IdentityServerCore.IdentityServerIdTokenValidator[0]
      Nonce in id_token does not match the nonce created for the login request - potential replay attack

**Comment by stuartbfs** [comment]
Is the nonce in the response received from IdentityServer4?

**Comment by chrismcv** [comment]
@vsavic - seeing something similar, did you ever make progress with this?

**Comment by vsavic** [comment]
Hey @chrismcv, sorry for delay... no I just switched to default [SS JWT auth](http://docs.servicestack.net/jwt-authprovider) to work with Identity Server.
